source/docs/software/wpilib-tools/robot-simulation/unit-testing.rst

```diff
@@ -0,0 +1,229 @@
+         };
+
+   .. group-tab:: C++ (Header)
+
+      .. code-block:: cpp
+
+         #include "subsystems/Intake.h"
+
+         void Intake::Deploy() {
+             piston.Set(frc::DoubleSolenoid::Value::kForward);
+         }
+
+         void Intake::Retract() {
+             piston.Set(frc::DoubleSolenoid::Value::kReverse);
+             motor.Set(0); // turn off the motor
+         }
+
+         void Intake::Activate(double speed) {
+             if (piston.Get() == frc::DoubleSolenoid::Value::kForward) {
+                 motor.Set(speed);
+             } else { // if piston isn't open, do nothing
+                 motor.Set(0);
+             }
+         }
+
+Writing Tests
+^^^^^^^^^^^^^
+
+Tests are placed inside the ``test`` source set: ``/src/test/java/`` and ``/src/test/cpp/`` for Java and C++ tests, respectively. In Java, each test class contains at least one test method marked with ``@org.junit.Test``, each method representing a test case. Additional methods for opening resources (such as our ``Intake`` object) before each test and closing them after are respectively marked with ``@org.junit.Before`` and ``@org.junit.After``. In C++, test fixture classes inheriting from ``testing::Test`` contain our subsystem and simulation hardware objects, and test methods are written using the ``TEST_F(testfixture, testname)`` macro. The ``SetUp()`` and ``TearDown()`` methods can be overridden in the test fixture class and will be run respectively before and after each test.
+
+Each test method should contain at least one *assertion* (``assert*()`` in Java or ``EXPECT_*()`` in C++). These assertions verify a condition at runtime and fail the test if the condition isn't met. If there is more than one assertion in a test method, the first failed assertion will crash the test - execution won't reach the later assertions.
+
+Both JUnit and GoogleTest have multiple assertion types, but the most common is equality: ``assertEquals(expected, actual)``/``EXPECT_EQ(expected, actual)``. When comparing numbers, a third parameter - ``delta``, the acceptable error, can be given. In JUnit (Java), these assertions are static methods and can be used without qualification by adding the static star import ``import static org.junit.Asssert.*``. In Google Test (C++), assertions are macros from the ``<gtest/gtest.h>`` header.
+
+.. note:: Comparison of floating-point values isn't accurate, so comparing them should be done with an acceptable error parameter (``DELTA``).
+
+.. tabs::
+   .. code-tab:: java
+
+      import static org.junit.Assert.*;
+
+      import edu.wpi.first.hal.HAL;
+      import edu.wpi.first.wpilibj.DoubleSolenoid;
+      import edu.wpi.first.wpilibj.simulation.DoubleSolenoidSim;
+      import edu.wpi.first.wpilibj.simulation.PWMSim;
+      import frc.robot.Constants.IntakeConstants;
+      import org.junit.*;
+
+      public class IntakeTest {
+        public static final double DELTA = 1e-2; // acceptable difference
```

Comment:
Not a big deal, but would `// deviation range` be more applicable?

Reply:
would `// acceptable deviation range` be ok in your opinion? to get the best of both options.